src/Lambertian.hpp

```diff
@@ -13,15 +15,9 @@
 
 class Lambertian: public Object {
   private:
-    static bool shaderProgramCompiled;
-    static unsigned int shaderProgram;
-    static unsigned int vertexShader;
-    static unsigned int fragmentShader;
+    GLShader shader;
```

Comment:
Can we have the shader be static? The reason being, switching shaders is probably the most costly operation so we want to do it as few times per frame as possible. I was thinking each "material" could have its own class extending `Object` that can manage binding all the things it needs to its shader when you draw it. When we have a bunch of objects in the scene, we'll sort them by material so we can switch shaders less

Reply:
I think representing the shader as essentially a global variable circumvents any automated cleanup. Instead, we could make the member a pointer, and enable and disable the shader in each draw loop (outside the `Object` draw method). I guess it's still a bit weird because the `Lambertian` class is still setting uniform values with raw OpenGL calls, but we can try moving this inside `GLShader` in another PR (I think that will be tricky due to the variable number of uniforms and types). Then the `GLShader` objects would be "materials", and `Object` subclasses would just represent a way to plumb data to a shader.

Thoughts?

Reply:
I'm ok with making the shader on the stack and making the member a pointer for now.

In terms of how we can avoid having a separate class to manage the shader for a given material: how do you feel about using a tuple to describe the information needing to be passed into a shader? We don't have introspection of structs, but you can get some of that using `template <class... Types> class std::tuple`. So we could do something like define the shader class as a template of a tuple of some type and then have a `draw` method for the shader take in a tuple of the same type. The implementation of that can maybe loop through the members and use `std::type_index` to switch on the type and use the corresponding gl function to bind it in. We can also probably `std::static_assert` that each type in the tuple is a type we know how to deal with, although idk what the syntax of that would be exactly, so I'm fine having a dynamic error thrown if we try to draw with a type we don't know how to handle.

Reply:
Alternatively, if we want to avoid metaprogramming hell, we can just make a struct with members for all the variables we need in a given material and then a method that takes in a shader reference and calls the appropriate binding functions. And then we can declare a `Shader<T>` for that struct `T`, and has a `draw(T)` method to pass in the struct of necessary variables. So `Lambertian` would just become an `Object<Lambertian>` which has a `Shader<Lambertian>`. Drawback here is having to manually write the bind calls in Lambertian still. I'm unsure that it's actually worth metaprogramming that part for this project since we likely won't be metaprogramming it for the fydp either, instead just doing code generation.

Reply:
I was considering using `std::tuple` for the uniform values, but I was concerned about enforcing access properly. I guess we could expose a `get<index>` function to extract the right type, but adding too much template magic when all of us aren't familiar with metaprogramming might be difficult to maintain.

I guess the second approach is a bit less magical, but then the templated type is essentially a bundle for providing uniforms, which seems a bit unintuitive IMO. We could also template `Shader` based on inputs, and also template `VertexBuffer` to ensure that the data being plumbed to the shader pipeline inputs is correctly formed.